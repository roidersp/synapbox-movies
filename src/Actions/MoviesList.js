import React from "react";

import gql from "graphql-tag";
import { Query } from "react-apollo";

import Grid from "@material-ui/core/Grid";

import MovieCard from "../Components/MovieCard";
import SearchBox from "../Components/SearchBox";

const GET_MOVIES = gql`
  {
    movies: itemsConnection(where: null, first: 24, after: null) {
      aggregate {
        count
      }
      edges {
        node {
          id
          title
          description
          image
          price
        }
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
  }
`;

const MoviesList = ({ action }) => (
  <Query query={GET_MOVIES}>
    {({ loading, error, data }) => {
      if (loading) return "Loading...";
      if (error) return `Error! ${error.message}`;

      const { edges: movies } = data.movies;

      return (
        <Grid container spacing={2}>
          <Grid item xs={12}>
            <SearchBox action={action} />
          </Grid>
          {movies.map(({ node: movie }, index) => (
            <MovieCard {...movie} key={index} action={action} />
          ))}
        </Grid>
      );
    }}
  </Query>
);

export default MoviesList;
