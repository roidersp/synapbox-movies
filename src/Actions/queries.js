import gql from "graphql-tag";

export const ADD_TO_CART = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) @client
  }
`;

export const DELETE_FROM_CART = gql`
  mutation deleteFromCart($id: ID!) {
    deleteFromCart(id: $id) @client
  }
`;

export const GET_CART_ITEMS = gql`
  query GetCartItems {
    cartItems @client
  }
`;

export const GET_MOVIE_ID = gql`
  {
    movieId @client
  }
`;

export const GET_MOVIE = gql`
  query($id: ID!) {
    movie: item(where: { id: $id }) {
      id
      title
      description
      largeImage
      price
    }
  }
`;

export const IS_IN_CART = gql`
  query isInCart($id: ID!) {
    isInCart(id: $id) @client
  }
`;

export const CART_IS_OPEN = gql`
  {
    cartIsOpen @client
    cartItems @client
  }
`;

export const CONFIRM_DELETE = gql`
  {
    dialogDelete @client {
      open
      movie {
        id
        title
        image
      }
    }
  }
`;

export const HANDLE_CONFIRM_DELETE = gql`
  {
    dialogDelete @client {
      open
      Movie
    }
  }
`;

export const GET_MOVIES = gql`
  {
    movies: itemsConnection(where: null, first: 16, after: null) {
      aggregate {
        count
      }
      edges {
        node {
          id
          title
          description
          image
          price
        }
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
  }
`;

export const GET_MOVIES_BY_ID = gql`
  query search($ids: [ID!]) {
    movies: items(where: { id_in: $ids }) {
      id
      title
      image
      price
    }
  }
`;

export const GET_SEARCH_MOVIES = gql`
  query search($searchValue: String!) {
    movies: itemsConnection(
      where: { title_contains: $searchValue }
      first: 5
      after: null
    ) {
      edges {
        node {
          id
          title
          description
          image
          price
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
`;
