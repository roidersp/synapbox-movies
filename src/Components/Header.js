import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import {
  AppBar,
  Toolbar,
  IconButton,
  Typography,
  Badge
} from "@material-ui/core";
import ShoppingCartIcon from "@material-ui/icons/ShoppingCart";

const useStyles = makeStyles(theme => ({
  grow: {
    flexGrow: 1
  },
  title: {
    [theme.breakpoints.down("sm")]: {
      fontSize: 14
    }
  }
}));

export default function PrimarySearchAppBar() {
  const classes = useStyles();

  return (
    <div className={classes.grow}>
      <AppBar position="static">
        <Toolbar>
          <Typography className={classes.title} variant="h6" noWrap>
            Synapbox Movies
          </Typography>
          <div className={classes.grow} />
          <div>
            <IconButton aria-label="show 4 new mails" color="inherit">
              <Badge badgeContent={0} color="secondary">
                <ShoppingCartIcon />
              </Badge>
            </IconButton>
          </div>
        </Toolbar>
      </AppBar>
    </div>
  );
}
