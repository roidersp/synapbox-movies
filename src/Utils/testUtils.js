import wait from "waait";
import { act } from "@testing-library/react";

export const actWait = async (amount = 0) => {
  await act(async () => {
    await wait(amount);
  });
};
