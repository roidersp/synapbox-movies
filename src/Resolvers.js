import gql from "graphql-tag";
import { GET_CART_ITEMS } from "./Actions/queries";

export const resolvers = {
  Mutation: {
    addToCart: (_, { id }, { cache }) => {
      const { cartItems } = cache.readQuery({ query: GET_CART_ITEMS });

      const data = {
        cartItems: !cartItems.includes(id) ? [...cartItems, id] : cartItems
      };

      cache.writeQuery({ query: GET_CART_ITEMS, data });
      return data.cartItems;
    },
    deleteFromCart: (_, { id }, { cache }) => {
      const { cartItems } = cache.readQuery({ query: GET_CART_ITEMS });

      const data = {
        cartItems: cartItems.includes(id)
          ? cartItems.filter(i => i !== id)
          : cartItems
      };

      cache.writeQuery({ query: GET_CART_ITEMS, data });
      return cartItems.includes(id);
    }
  }
};

export const typeDefs = `
  extend type Mutation {
    addToCart(id: ID!)
    deleteFromCart(id: ID!)
  }
`;
