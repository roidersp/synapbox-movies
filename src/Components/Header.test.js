import React from "react";
import { MockedProvider } from "@apollo/react-testing";
import { InMemoryCache } from "apollo-cache-inmemory";
import { actWait } from "../Utils/testUtils";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { GET_CART_ITEMS, CART_IS_OPEN } from "../Actions/queries";
import { resolvers, typeDefs } from "../Resolvers";

import Header from "./Header";

afterEach(cleanup);

const mocks = [
  {
    request: { query: GET_CART_ITEMS },
    result: {
      data: {
        cartItems: ["ck49086bqsnup0b09d14atpb2", "ck4909pm29tze0b00xnb2nr6p"]
      }
    }
  }
];

describe("Header testing", () => {
  it("snapshot", async () => {
    const cache = new InMemoryCache();

    cache.writeData({
      data: {
        cartItems: [],
        cartIsOpen: false
      }
    });

    const Wrapper = render(
      <MockedProvider
        mocks={mocks}
        addTypename={false}
        resolvers={{}}
        cache={cache}
      >
        <Header />
      </MockedProvider>
    );
    await actWait();
    expect(Wrapper).toMatchSnapshot();
  });
  it("Should show title", async () => {
    const cache = new InMemoryCache();

    cache.writeData({
      data: {
        cartItems: [],
        cartIsOpen: false
      }
    });

    const { queryByText } = render(
      <MockedProvider
        mocks={mocks}
        addTypename={false}
        resolvers={{}}
        cache={cache}
      >
        <Header />
      </MockedProvider>
    );

    await actWait();

    expect(queryByText("Synapbox Movies")).toBeTruthy();
  });

  it("Click open cart", async () => {
    const cache = new InMemoryCache();

    cache.writeData({
      data: {
        cartItems: [],
        cartIsOpen: false
      }
    });

    const { getByTestId } = render(
      <MockedProvider
        mocks={mocks}
        addTypename={false}
        resolvers={{}}
        cache={cache}
      >
        <Header />
      </MockedProvider>
    );

    fireEvent.click(getByTestId("openCart"));

    await actWait();

    const { cartIsOpen } = cache.readQuery({ query: CART_IS_OPEN });

    expect(cartIsOpen).toBeTruthy();
  });
});
