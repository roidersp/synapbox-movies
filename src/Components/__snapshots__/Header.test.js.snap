// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header testing snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-grow-1"
      >
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
          >
            <h6
              class="MuiTypography-root makeStyles-title-2 MuiTypography-h6 MuiTypography-noWrap"
            >
              Synapbox Movies
            </h6>
            <div
              class="makeStyles-grow-1"
            />
            <div>
              <button
                aria-label="0 peliculas en el carrito"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
                data-testid="openCart"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <span
                    class="MuiBadge-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
                      />
                    </svg>
                    <span
                      class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorSecondary MuiBadge-invisible"
                    >
                      0
                    </span>
                  </span>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </header>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-grow-1"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <h6
            class="MuiTypography-root makeStyles-title-2 MuiTypography-h6 MuiTypography-noWrap"
          >
            Synapbox Movies
          </h6>
          <div
            class="makeStyles-grow-1"
          />
          <div>
            <button
              aria-label="0 peliculas en el carrito"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
              data-testid="openCart"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <span
                  class="MuiBadge-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorSecondary MuiBadge-invisible"
                  >
                    0
                  </span>
                </span>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </header>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
